chain-signatures/node/src/mesh/mod.rs

```diff
@@ -29,89 +31,61 @@ impl Options {
-    pub fn active_participants(&self) -> &Participants {
-        &self.active_participants
+            state: state.clone(),
+        };
+        (mesh, state)
     }
 
-    /// Potential participants that are active at the beginning of each protocol loop. This will
-    /// be empty if not in resharing state for the protocol
-    pub fn active_potential_participants(&self) -> &Participants {
-        &self.active_potential_participants
+    async fn ping(&mut self) {
+        let mut mesh_state = self.state.write().await;
+        *mesh_state = MeshState {
+            active_participants: self.connections.ping().await,
+            active_potential_participants: self.connections.ping_potential().await,
+            potential_participants: self.connections.potential_participants().await,
+            stable_participants: self.connections.stable_participants().await,
+        };
     }
 
-    /// Get all pontential participants, but they may not necessarily be active.
-    pub async fn potential_participants(&self) -> Participants {
-        self.connections.potential_participants().await
-    }
-
-    pub fn all_active_participants(&self) -> Participants {
-        let mut participants = self.active_participants.clone();
-        let active = self.active_potential_participants.keys_vec();
-        tracing::info!(?active, "Getting potentially active participants");
-        for (participant, info) in self.active_potential_participants.iter() {
-            if !participants.contains_key(participant) {
-                participants.insert(participant, info.clone());
-            }
-        }
-        participants
-    }
-
-    /// Get active participants that have a stable connection. This is useful for arbitrary metrics to
-    /// say whether or not a node is stable, such as a node being on track with the latest block height.
-    ///
-    /// NOTE: ping() will also ping our own node since everyone, including us, is stored in the contract state
-    /// or `ProtocolState`, which eventually leads to checking our own stablitity.
-    pub async fn stable_participants(&self) -> Participants {
-        let mut stable = Participants::default();
-        for (participant, info) in self.active_participants().iter() {
-            if self.connections.is_participant_stable(participant).await {
-                stable.insert(participant, info.clone());
+    pub async fn run(
+        mut self,
+        contract_state: Arc<RwLock<Option<ProtocolState>>>,
+    ) -> anyhow::Result<()> {
+        let mut last_pinged = Instant::now();
+        loop {
+            if last_pinged.elapsed() > Duration::from_millis(300) {
+                if let Some(state) = contract_state.read().await.clone() {
+                    self.connections.establish_participants(&state).await;
+                    self.ping().await;
+                    last_pinged = Instant::now();
+                }
             }
+            tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
no need to check every 300ms, rather just have sleep and we can get rid of maintaining `last_pinged` as well. Also no need for state clone.
```suggestion
            {
                let state = contract_state.read().await;
                if let Some(state) =  {
                    self.connections.establish_participants(&*state).await;
                }
            }
            tokio::time::sleep(Duration::from_millis(300)).await;
```